### ISSUE TITLE
Unscale gradients before gradient clipping

### ISSUE THREAD

**Issue by ivan-chai** [catalyst-team/catalyst#1319]
## 🐛 Bug Report
OptimizerCallback doesn't handle AMP correctly. Gradients need to be unscaled before clipping.

Original documentation is here:

[https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-clipping](url)

Catalyst 21.09 code:
```
    def on_batch_end(self, runner: "IRunner"):
        """Event handler."""
        if runner.is_train_loader:
            self._accumulation_counter += 1
            need_gradient_step = self._accumulation_counter % self.accumulation_steps == 0

            loss = runner.batch_metrics[self.metric_key]
            runner.engine.backward_loss(loss, self.model, self.optimizer)

            if self.grad_clip_fn is not None:
                self.grad_clip_fn(self.model.parameters())

            if need_gradient_step:
                runner.engine.optimizer_step(loss, self.model, self.optimizer)
                runner.engine.zero_grad(loss, self.model, self.optimizer)

        runner.batch_metrics.update(self._get_lr_momentum_stats())
```

**Comment by Scitator** [comment]
That's interesting 👍

**Comment by Scitator** [comment]
working :in-progress:
